src/app/src/main/kotlin/org/vaccineimpact/api/app/controllers/PasswordController.kt

```diff
@@ -1,26 +1,69 @@
 package org.vaccineimpact.api.app.controllers
 
+import org.vaccineimpact.api.OneTimeAction
 import org.vaccineimpact.api.app.ActionContext
+import org.vaccineimpact.api.app.OneTimeLinkActionContext
 import org.vaccineimpact.api.app.controllers.endpoints.oneRepoEndpoint
 import org.vaccineimpact.api.app.controllers.endpoints.secured
+import org.vaccineimpact.api.app.errors.MissingRequiredParameterError
 import org.vaccineimpact.api.app.models.SetPassword
 import org.vaccineimpact.api.app.postData
 import org.vaccineimpact.api.app.repositories.Repositories
 import org.vaccineimpact.api.app.repositories.UserRepository
+import org.vaccineimpact.api.emails.EmailManager
+import org.vaccineimpact.api.emails.PasswordSetEmail
+import org.vaccineimpact.api.emails.getEmailManager
 import spark.route.HttpMethod
 
-class PasswordController(context: ControllerContext) : AbstractController(context)
+class PasswordController(
+        context: ControllerContext,
+        val emailManager: EmailManager = getEmailManager()
+) : AbstractController(context)
 {
     override val urlComponent = "/password"
 
     override fun endpoints(repos: Repositories) = listOf(
-            oneRepoEndpoint("/set/", this::setPassword, repos.user, HttpMethod.post).secured()
+            oneRepoEndpoint("/set/", this::setPassword, repos.user, HttpMethod.post).secured(),
+            oneRepoEndpoint("/request_link/", this::requestLink, repos.user, HttpMethod.post)
     )
 
+    private val tokenRepoFactory by lazy { context.repositories.token }
+
     fun setPassword(context: ActionContext, repo: UserRepository): String
     {
+        return setPasswordForUser(context, repo, context.username!!)
+    }
+    fun setPasswordForUser(context: ActionContext, repo: UserRepository, username: String): String
+    {
         val password = context.postData<SetPassword>().password
-        repo.setPassword(context.username!!, password)
+        repo.setPassword(username, password)
+        return okayResponse()
+    }
+
+    fun requestLink(context: ActionContext, repo: UserRepository): String
+    {
+        val address = context.queryParams("email")
+                ?: throw MissingRequiredParameterError("email")
+        val user = repo.getMontaguUserByEmail(address)
+        if (user != null)
+        {
+            val token = getSetPasswordToken(user.username, context)
```

Comment:
The architecture is getting a bit confusing here - really the endpoint that uses this method is not a 'one repo' endpoint because it depends on both the user and the token repository, the latter being accessed directly from the controller context, the former passed in to the method.

The failing unit tests are because the token repo has not been mocked on the mock controller context, but I can see why that was missed because there isn't a clear dependency injection pattern! Overall I do think this demonstrates the limitation of the 'one repo endpoint' and that we should use a consistent depedency injection pattern.

 I'm tempted to say given time constraints   just fix the tests and open a ticket for refactoring, but the danger is then that the refactor will never be a top priority. Let's discuss in person.